DeviceAdapters/Zaber/WdiAutofocus.cpp

```diff
@@ -0,0 +1,346 @@
+}
+
+WdiAutofocus::~WdiAutofocus()
+{
+	this->LogMessage("WdiAutofocus::~WdiAutofocus\n", true);
+	Shutdown();
+}
+
+void WdiAutofocus::GetName(char* name) const
+{
+	CDeviceUtils::CopyLimitedString(name, g_WdiAutofocusDescription);
+}
+
+int WdiAutofocus::Initialize()
+{
+	if (initialized_)
+	{
+		return DEVICE_OK;
+	}
+
+	core_ = GetCoreCallback();
+
+	this->LogMessage("WdiAutofocus::Initialize\n", true);
+
+	auto pAct = new CPropertyAction(this, &WdiAutofocus::LimitMinGetSet);
+	CreateFloatProperty("Limit Min [mm]", limitMin_, false, pAct);
+	pAct = new CPropertyAction(this, &WdiAutofocus::LimitMaxGetSet);
+	CreateFloatProperty("Limit Max [mm]", limitMax_, false, pAct);
+
+	auto ret = UpdateStatus();
+	if (ret == DEVICE_OK)
+	{
+		initialized_ = true;
+	}
+
+	return ret;
+}
+
+int WdiAutofocus::Shutdown()
+{
+	this->LogMessage("WdiAutofocus::Shutdown\n", true);
+
+	if (initialized_)
+	{
+		initialized_ = false;
+	}
+
+	return DEVICE_OK;
+}
+
+bool WdiAutofocus::Busy()
+{
+	this->LogMessage("WdiAutofocus::Busy\n", true);
+
+	bool busy = false;
+	auto ret = handleException([&]() {
+		ensureConnected();
+		busy = axis_.isBusy();
+	});
+	return ret == DEVICE_OK && busy;
+}
+
+int WdiAutofocus::PortGetSet(MM::PropertyBase* pProp, MM::ActionType eAct)
+{
+	ostringstream os;
+	os << "WdiAutofocus::PortGetSet(" << pProp << ", " << eAct << ")\n";
+	this->LogMessage(os.str().c_str(), false);
+
+	if (eAct == MM::BeforeGet)
+	{
+		pProp->Set(port_.c_str());
+	}
+	else if (eAct == MM::AfterSet)
+	{
+		if (initialized_)
+		{
+			resetConnection();
+		}
+
+		pProp->Get(port_);
+	}
+
+	return DEVICE_OK;
+}
+
+int WdiAutofocus::FocusAddressGetSet(MM::PropertyBase* pProp, MM::ActionType eAct)
+{
+	this->LogMessage("WdiAutofocus::FocusAddressGetSet\n", true);
+
+	if (eAct == MM::AfterSet)
+	{
+		if (initialized_)
+		{
+			resetConnection();
+		}
+
+		pProp->Get(focusAddress_);
+	}
+	else if (eAct == MM::BeforeGet)
+	{
+		pProp->Set(focusAddress_);
+	}
+
+	return DEVICE_OK;
+}
+
+int WdiAutofocus::FocusAxisGetSet(MM::PropertyBase* pProp, MM::ActionType eAct)
+{
+	this->LogMessage("WdiAutofocus::FocusAxisGetSet\n", true);
+
+	if (eAct == MM::AfterSet)
+	{
+		if (initialized_)
+		{
+			resetConnection();
+		}
+
+		pProp->Get(focusAxis_);
+	}
+	else if (eAct == MM::BeforeGet)
+	{
+		pProp->Set(focusAxis_);
+	}
+
+	return DEVICE_OK;
+}
+
+int WdiAutofocus::LimitMinGetSet(MM::PropertyBase* pProp, MM::ActionType eAct)
+{
+	this->LogMessage("ObjectiveChanger::LimitMinGetSet\n", true);
+
+	if (eAct == MM::AfterSet)
+	{
+		double newLimit;
+		pProp->Get(newLimit);
+		auto update = limitMin_ != newLimit;
+		limitMin_ = newLimit;
+		if (update) {
+			return handleException([=]() {
+				ensureConnected();
+				axis_.getSettings().set("motion.tracking.limit.min", limitMin_ * WdiAutofocus_::xLdaNativePerMm);
+			});
+		}
+	}
+	else if (eAct == MM::BeforeGet)
+	{
+		if (initialized_)
+		{
+			int ret = handleException([=]() {
```

Comment:
clang-tidy doesn't like using `=` to capture `this`

```
error: lambdas that implicitly capture 'this' should not specify a by-value capture default [cppcoreguidelines-misleading-capture-default-by-value,-warnings-as-errors]
  473 |     return zml_util::map_vec<int,Device>(response.getDevices(), [=](const int& i) { return Device(*this, i); });
      |                                                                  ^
      |                                                                  this
```

Reply:
It's funny because `[this]` also captures the `this` ptr by value, so both achieve the same result. I just remember this linter rule coming up when working on replacing RxCpp so thought I'd point it out

Reply:
fixed